nodejs/helpers.js

```diff
@@ -0,0 +1,172 @@
+      this.regularClient = null
+      this.csfleClient = null
+    }
+
+    /**
+     * Creates a unique, partial index in the key vault collection
+     * on the ``keyAltNames`` field.
+     *
+     * @param {MongoClient} client
+     */
+    async ensureUniqueIndexOnKeyVault(client) {
+      try {
+        await client
+          .db(this.keyDB)
+          .collection(this.keyColl)
+          .createIndex("keyAltNames", {
+            unique: true,
+            partialFilterExpression: {
+              keyAltNames: {
+                $exists: true
+              }
+            }
+          })
+      } catch (e) {
+        console.error(e)
+        process.exit(1)
+      }
+    }
+
+    /**
+     * In the guide, https://docs.mongodb.com/ecosystem/use-cases/client-side-field-level-encryption-guide/,
+     * we create the data key and then show that it is created by
+     * retreiving it using a findOne query. Here, in implementation, we only
+     * create the key if it doesn't already exist, ensuring we only have one
+     * local data key.
+     */
+    async findOrCreateDataKey() {
+      const client = new MongoClient(this.connectionString, {
+        useNewUrlParser: true,
+        useUnifiedTopology: true
+      })
+      await client.connect()
+
+      const encryption = new ClientEncryption(client, {
+        keyVaultNamespace: this.keyVaultNamespace,
+        kmsProviders: this.kmsProviders
+      })
+
+      await this.ensureUniqueIndexOnKeyVault(client)
+
+      let dataKey = await client
+        .db(this.keyDB)
+        .collection(this.keyColl)
+        .findOne({ keyAltNames: { $in: [this.keyAltNames] } })
+
+      if (dataKey === null) {
+        dataKey = await encryption.createDataKey("local", {
+          keyAltNames: [this.keyAltNames]
+        })
+      }
+
+      await client.close()
+      return dataKey
+    }
+
+    async getRegularClient() {
+      const client = new MongoClient(this.connectionString, {
+        useNewUrlParser: true,
+        useUnifiedTopology: true
+      })
+      return await client.connect()
+    }
+
+    async getCsfleEnabledClient(schemaMap = null) {
+      if (schemaMap === null) {
+        throw new Error(
+          "schemaMap is a required argument. Build it using the CsfleHelper.createJsonSchemaMap method"
+        )
+      }
+      const client = new MongoClient(this.connectionString, {
+        useNewUrlParser: true,
+        useUnifiedTopology: true,
+        monitorCommands: true,
+        autoEncryption: {
+          keyVaultNamespace: this.keyVaultNamespace,
+          kmsProviders: this.kmsProviders,
+          schemaMap
+        }
+      })
+      return await client.connect()
+    }
+
+    createJsonSchemaMap(dataKey = null) {
+      if (dataKey === null) {
+        throw new Error(
+          "dataKey is a required argument. Ensure you've defined it in clients.js"
+        )
+      }
+      return {
+        "medicalRecords.patients": {
```

Comment:
thought: would it make sense to store the majority of your JSON schema in a separate file? I know this has the `keyId` added to it, but maybe it would make these files easier to reason about if you weren't staring at a bunch of JSON.

Reply:
It most likely would! We've done this mainly for consistency. I'll speak with the team to see if we want to do this.